drgn_tools/slabinfo.py

```diff
@@ -0,0 +1,310 @@
+                "slab_list",
+            ):
+                nrobj = page.objects.value_()
+                nrinuse = page.inuse.value_()
+                x = nrobj - nrinuse
+                node_total_use += nrinuse
+                nr_free += x
+        except LookupError:
+            for page in list_for_each_entry(
+                "struct page", kmem_cache_node.partial.address_of_(), "lru"
+            ):
+                nrobj = page.objects.value_()
+                nrinuse = page.inuse.value_()
+                x = nrobj - nrinuse
+                node_total_use += nrinuse
+                nr_free += x
+
+        for cpuid in for_each_online_cpu(prog):
+            try:
+                per_cpu_slab = per_cpu_ptr(s.cpu_slab, cpuid)
+            except (AttributeError, drgn.ObjectAbsentError):
+                per_cpu_slab = 0
+            else:
+                cpu_slab_ptr = per_cpu_slab.page
+                if not cpu_slab_ptr:
+                    continue
+                cpu_per_node += 1
+
+        full_slabs = nr_slabs - cpu_per_node - nr_partial
+
+    total_inuse += nr_total_objs
+    free_objects += nr_free
+    total_slabs += nr_partial
+    total_slabs += full_slabs
+    inuse = total_inuse - free_objects
+    output.append(
+        [
+            ad,
+            int(s.object_size),
+            inuse,
+            total_inuse,
+            total_slabs,
+            f"{ssize}k",
+            s.name.string_().decode("utf-8"),
+        ]
+    )
+
+
+def freelist_dereference_swab(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Freelist traversal with swab
+    """
+    freelist_random = slab_cache.random.value_()
+    ulong_size = sizeof(prog.type("unsigned long"))
+    return (
+        prog.read_word(ptr_addr)
+        ^ freelist_random
+        ^ int.from_bytes(ptr_addr.to_bytes(ulong_size, "little"), "big")
+    )
+
+
+def freelist_dereference_noswab(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Freelist traversal with no swab
+    """
+    freelist_random = slab_cache.random.value_()
+    return prog.read_word(ptr_addr) ^ freelist_random ^ ptr_addr
+
+
+def try_hardened_freelist_dereference(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Segregate the free list traverse based on CONFIG_SLAB_FREELIST_HARDENED
+    """
+    try:
+        slab_cache.random.value_()
+
+    except AttributeError:
+        slab_freelist_hardened = 0
+        needswab = 0
+
+    else:
+        slab_freelist_hardened = 1
+        needswab = 1
+
+    if slab_freelist_hardened and needswab:
+        result = freelist_dereference_swab(prog, slab_cache, ptr_addr)
+    if not slab_freelist_hardened and not needswab:
+        try:
+            result = prog.read_word(ptr_addr)
+        except (AttributeError, drgn.ObjectAbsentError):
+            result = 0
+
+    return result
```

Comment:
1. As far as I can tell, this ignores the `freelist_dereference_noswab()` case.
2. It looks like these parts of the code are adapted from the drgn slab code. In the drgn slab helpers there are comments regarding the different commits that cause us to need each of the variants. The drgn slab helpers also support `struct slab` rather than `struct page`, which we will need for UEK8.
3. This doesn't end up caching the `freelist_dereference()` choice, which means we end up redoing the checks for every time we dereference a freelist pointer.

I think we could solve all three, by simply using the `drgn.helpers.linux.slab._SlabCacheHelperSlub` class. You'll need to read the code to find it, but it provides most of this stuff in a way that's more compatible, more efficient, and better tested.

Essentially, inside of the `kmem_cache_info_slub()` function, you would do something like this:

```python
helper = _SlabCacheHelperSlub(cache)
```

On creation, `helper` will have a function `helper._slub_get_freelist()` which you can pass a freelist pointer into, and a set, and it will add each element of the freelist in, using the correct method.

I know that the downside here is that the `_SlabCacheHelperSlub` class is not a public API of drgn. But for one thing, each drgn-tools release is specific to a drgn release, so we know we won't get suddenly broken. For another, I think it would be worthwhile to try to make some of that API public, that's something we can discuss with Omar a bit.